EIPS/eip-223.md

```diff
@@ -0,0 +1,385 @@
+```
+
+This function must transfer tokens and invoke the function `tokenReceived (address, uint256, bytes)` in `_to`, if `_to` is a contract. If the `tokenReceived` function is not implemented in `_to` (recipient contract), then the transaction must fail and the transfer of tokens must not occur. 
+If `_to` is an externally owned address, then the transaction must be sent without executing `tokenReceived` in `_to`.
+ `_data` can be attached to this token transaction and it will stay in blockchain forever (requires more gas). `_data` can be empty.
+
+NOTE: A possible way to check whether the `_to` is a contract or an address is to assemble the code of `_to`. If there is no code in `_to`, then this is an externally owned address, otherwise it's a contract.
+
+### Contract that is intended to receive [ERC-223](./eip-223.md) tokens
+
+```solidity
+function tokenReceived(address _from, uint _value, bytes calldata _data)
+```
+
+A function for handling token transfers, which is called from the token contract, when a token holder sends tokens. `_from` is the address of the sender of the token, `_value` is the amount of incoming tokens, and `_data` is attached data similar to `msg.data` of Ether transactions. It works by analogy with the fallback function of Ether transactions and returns nothing.
+
+NOTE: `msg.sender` will be a token-contract inside the `tokenReceived` function. It may be important to filter which tokens are sent (by token-contract address). The token sender (the person who initiated the token transaction) will be `_from` inside the `tokenReceived` function.
+
+IMPORTANT: This function must be named `tokenReceived` and take parameters `address`, `uint256`, `bytes` to match the function signature `0x8943ec02`.
+
+### Events
+
+#### `Transfer`
+
+```solidity
+event Transfer(address indexed _from, address indexed _to, uint256 _value)
+```
+
+Triggered when tokens are transferred. Compatible with ERC-20 `Transfer` event. The functionality of the event is similar to `Transfer` event of ERC-20.
+
+
+## Rationale
+
+This standard introduces a communication model by enforcing the `transfer` to execute a handler function in the destination address. This is an important security consideration as it is required that the receiver explicitly implements the token handling function. In cases where the receiver does not implements such function the transfer MUST be reverted.
+
+This standard sticks to the push transaction model where the transfer of assets is initiated on the senders side and handled on the receivers side. As the result, ERC-223 transfers are more gas-efficient while dealing with depositing to contracts as ERC-223 tokens can be deposited with just one transaction while ERC-20 tokens require at least two calls (one for `approve` and the second that will invoke `transferFrom`).
+
+- [ERC-20](./eip-20.md) deposit: `approve` ~53K gas, `transferFrom` ~80K gas
+
+- ERC-223 deposit: `transfer` and handling on the receivers side ~46K gas
```

Comment:
How does the deposit + call use less gas than an approve? I feel like this gas usage might be outdated

Reply:
I have performed this tests https://github.com/Dexaran/Token_Deposits_GAS_testing using REMIX, solidity 0.8.19, optimization 200 runs

Updated the values:

-  approve >> 53K
- deposit with transferFrom >> 75K
- ERC223 deposit via transfer >> 46K

Reply:
Your transferFrom function is very unoptimized, with 7 total internal contract calls. Your ERC-223 contract has none. This skews the results a lot.